genie_registry/clinical.py

```diff
@@ -472,6 +472,68 @@ def process_steps(
         newClinicalDf.to_csv(newPath, sep="\t", index=False)
         return newPath
 
+    def _validate_oncotree_code_mapping(
+        self: "Clinical", clinicaldf: pd.DataFrame, oncotree_mapping: pd.DataFrame
+    ) -> pd.Index:
+        """Checks that the oncotree codes in the input clinical
+        data is a valid oncotree code from the official oncotree site
+
+        Args:
+            clinicaldf (pd.DataFrame): clinical input data to validate
+            oncotree_mapping (pd.DataFrame): table of official oncotree
+                mappings
+
+        Returns:
+            pd.Index: row indices of unmapped oncotree codes in the
+            input clinical data
+        """
+        # Make oncotree codes uppercase (SpCC/SPCC)
+        clinicaldf["ONCOTREE_CODE"] = (
+            clinicaldf["ONCOTREE_CODE"].astype(str).str.upper()
+        )
+
+        unmapped_oncotrees = clinicaldf[
+            (clinicaldf["ONCOTREE_CODE"] != "UNKNOWN")
+            & ~(clinicaldf["ONCOTREE_CODE"].isin(oncotree_mapping["ONCOTREE_CODE"]))
+        ]
+        return unmapped_oncotrees.index
+
+    def _validate_oncotree_code_mapping_message(
+        self: "Clinical",
+        clinicaldf: pd.DataFrame,
+        unmapped_oncotree_indices: pd.DataFrame,
+    ) -> Tuple[str, str]:
+        """This function returns the error and warning messages
+        if the input clinical data has row indices with unmapped
+        oncotree codes
+
+        Args:
+            clinicaldf (pd.DataFrame): input clinical data
+            unmapped_oncotree_indices (pd.DataFrame): row indices of the
+                input clinical data with unmapped oncotree codes
+
+        Returns:
+            Tuple[str, str]: error message that tells you how many
+                samples AND the unique unmapped oncotree codes that your
+                input clinical data has
+        """
+        errors = ""
+        warnings = ""
+        if len(unmapped_oncotree_indices) > 0:
+            # sort the unique unmapped oncotree codes
+            unmapped_oncotree_codes = sorted(
+                set(clinicaldf.loc[unmapped_oncotree_indices]["ONCOTREE_CODE"])
+            )
+            errors = (
+                "Sample Clinical File: Please double check that all your "
+                "ONCOTREE CODES exist in the mapping. You have {} samples "
+                "that don't map. These are the codes that "
+                "don't map: {}\n".format(
+                    len(unmapped_oncotree_indices), ",".join(unmapped_oncotree_codes)
+                )
+            )
+        return errors, warnings
```

Comment:
Nit: at some point, I'd love for us to not return a tuple, but a namedtuple or something.  This has always bothered me since the beginning, but its a pretty significant code change.

Reply:
Created a ticket for this: https://sagebionetworks.jira.com/browse/GEN-1265 so it's on our radar